### PR TITLE
Deal with duplicate status updates in retrying_executor

### PR DESCRIPTION
**Pull request by sagar8192**

### CODE REVIEW

File: task_processing/plugins/mesos/retrying_executor.py [hunk 1]
```diff
@@ -36,34 +36,49 @@ def event_with_retries(self, event):
...
         while True:
             while not self.src_queue.empty():
                 e = self.src_queue.get()
+                # This is to remove trailing '-retry*'
+                original_task_id = '-'.join([item for item in
+                                             e.task_id.split('-')[:-1]])
+
+                # Check if the update is for current attempt. Discard if
+                # it is not.
+                if not self._is_update_duplicate(e, original_task_id):
```
Comment (keymone) [inline task_processing/plugins/mesos/retrying_executor.py]: function name doesn't really reflect the purpose, could you change to `is_current_attempt` or something?

File: task_processing/plugins/mesos/retrying_executor.py [hunk 2]
```diff
@@ -36,34 +36,49 @@ def event_with_retries(self, event):
...
+                    continue
+
+                # Set the task id back to original task_id
+                e = self._decode_event(e, original_task_id)
```
Comment (keymone) [inline task_processing/plugins/mesos/retrying_executor.py]: another function name nitpick, maybe `restore_task_id`?

File: task_processing/plugins/mesos/retrying_executor.py [hunk 3]
```diff
@@ -91,8 +106,8 @@ def run(self, task_config):
         if task_config.task_id not in self.task_retries:
             with self.task_retries_lock:
                 self.task_retries = self.task_retries.set(
-                    task_config.task_id, self.retries)
-        self.executor.run(task_config)
+                    task_config.task_id, 1)
+        self.executor.run(self._get_encoded_task_config(task_config))
```
Comment (keymone) [inline task_processing/plugins/mesos/retrying_executor.py]: and one more function name nitpick: `task_config_with_retry` ?

File: task_processing/plugins/mesos/retrying_executor.py [hunk 4]
```diff
@@ -36,34 +36,49 @@ def event_with_retries(self, event):
...
             )
         )
 
     def retry(self, event):
-        current_retries = self.task_retries.get(event.task_id, -1)
-        if current_retries <= 0:
+        with self.task_retries_lock:
+            current_retry_attempt = self.task_retries[event.task_id]
+
+        if current_retry_attempt == self.retries:
```
[the remark this answers is not in the document]
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Good call. Fixed.

File: task_processing/plugins/mesos/retrying_executor.py (hunk 1 above)
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Sure. Fixed.

File: task_processing/plugins/mesos/retrying_executor.py (hunk 2 above)
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Fixed.

File: task_processing/plugins/mesos/retrying_executor.py (hunk 3 above)
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Fixed.

File: task_processing/plugins/mesos/retrying_executor.py [hunk 5]
```diff
@@ -108,3 +123,38 @@ def stop(self):
...
+        )
+
+    def _is_update_duplicate(self, e, original_task_id):
+        retry_suffix = '-'.join([item for item in
+                                 e.task_id.split('-')[-1:]])
+
+        # This is to extract retry attempt from retry_suffix
+        # eg: if retry_suffix= 'retry2', then attempt==2
+        attempt = int(retry_suffix[5:])
+        with self.task_retries_lock:
```
[the remark this answers is not in the document]
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Yeah, you are right. Fixed.

File: task_processing/plugins/mesos/retrying_executor.py [hunk 6]
```diff
@@ -36,34 +36,49 @@ def event_with_retries(self, event):
...
-                1 + self.retries - self.task_retries.get(event.task_id, -1),
+                self.task_retries[event.task_id],
                 self.retries
             )
         )
 
     def retry(self, event):
-        current_retries = self.task_retries.get(event.task_id, -1)
-        if current_retries <= 0:
+        with self.task_retries_lock:
```
[the remark this answers is not in the document]
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: Yeah, you are right. Fixed.

File: task_processing/plugins/mesos/retrying_executor.py [hunk 7]
```diff
@@ -108,3 +123,38 @@ def stop(self):
...
+    def _get_encoded_task_config(self, task_config):
+        return task_config.set(uuid='{id}-retry{attempt}'.format(
+            id=task_config.uuid,
+            attempt=self.task_retries[task_config.task_id]
+        ))
+
+    def _decode_event(self, e, original_task_id):
+        # Fix task_id references
+        mesos_status = e.raw
+        mesos_status.task_id.value = original_task_id
```
[the remark this answers is not in the document]
Comment (sagar8192) [reply task_processing/plugins/mesos/retrying_executor.py]: If users are using the raw event for something, then they will be a discrepancy between task_id in the event and task_id in status update. I am modifying the raw event to avoid this. But, I don't have a strong opinion about this. Accepted your suggestion for now. If any user complains about this, then we can restore the original task_id.

File: task_processing/plugins/mesos/mesos_executor.py [hunk 8]
```diff
@@ -41,7 +41,7 @@ def __invariant__(conf):
         return ('image' in conf if conf.containerizer == 'DOCKER' else True,
                 'Image required for chosen containerizer')
 
-    uuid = field(type=uuid.UUID, initial=uuid.uuid4)
+    uuid = field(type=str, initial=str(uuid.uuid4()))
```
Comment (keymone) [inline task_processing/plugins/mesos/mesos_executor.py]: what's the reason behind this change?

Comment (sagar8192) [reply task_processing/plugins/mesos/mesos_executor.py]: uuid for returns an object of type UUID. We cannot append a string to it.

Comment (keymone) [reply task_processing/plugins/mesos/mesos_executor.py]: right, so it shouldn't be called uuid then

**Comment by keymone** [comment]
do check why the tests are failing before merging
